### DISCUSSION TITLE
How to use @profile in Pluto?

### DISCUSSION THREAD

**Discussion by gdalle** [fonsp/Pluto.jl#1431] [Questions about using Pluto]
When I profile my own functions in Pluto, I often end up with three external methods (`task_done_hook`, `wait` and `pop_task` from `task.jl` in Base) taking most of the CPU time. See the example below, taken from my [course material](https://gdalle.github.io/IntroJulia/notebooks/efficiency.jl.html).
![Screenshot from 2021-08-30 12-16-43](https://user-images.githubusercontent.com/22795598/131324946-f32b1eb9-a381-45d1-a4a7-233d90054505.png)
From what I understand, it could be [linked to multiprocessing](https://discourse.julialang.org/t/profiling-code-reveals-that-most-time-is-used-calling-3-functions-in-task-jl/30751/6), more precisely the fact that Pluto launches each notebook on a separate worker (see #300). Is there a workaround to get precise profiling without the multiprocessing stuff?

**Comment by gdalle** [comment]
I found a workaround that doesn't act on the profiling itself but on the generated FlameGraph. It selects the subtree of the FlameGraph where the code is actually run, and discards the time spent waiting on the designated worker. I think it may work as long as the code we profile doesn't use Pluto itself.

```julia
using AbstractTrees, FlameGraphs

function select_pluto_subgraph(g)
    for n in PreOrderDFS(g)
        if (
            endswith(string(n.data.sf.file), "PlutoRunner.jl") &&
            endswith(string(n.data.sf.func), "run_inside_trycatch")
        )
            return n
        end
    end
end
```
From there, if we want to profile a function `myfunc`, all we have to do is run
```julia
using Profile, ProfileSVG

Profile.clear()
@profile f(args...)
ProfileSVG.view(select_pluto_subgraph(flamegraph()))
```
@fonsp does this look right to you? Would it be interesting to wrap this in a macro, say `@plutoprofile`? Would it belong in Pluto itself or rather in a mini-package outside?

**Reply by VarLad** [reply]
This looks awesome! Thanks for putting in the work!

**Comment by gdalle** [comment]
See https://github.com/gdalle/PlutoProfile.jl

**Comment by fonsp** [comment]
Thanks for pointing this out and for investigating the issue!

Note the difference: **Multiprocessing** (using [`Distributed`](https://docs.julialang.org/en/v1/manual/distributed-computing/), different OS process for each notebook) is not the same as [**multitasking** (`@async`)](https://docs.julialang.org/en/v1/manual/asynchronous-programming/), which is again different from [**multithreading** (running tasks in parallel within the same process)](https://docs.julialang.org/en/v1/base/multi-threading/).

But they are related: We use multiprocessing to manage different notebook processing, which (I believe) spawns tasks on the notebook process to execute actions in, including code execution. We do not use multithreading in the notebook process (I think).

**Reply by gdalle** [reply]
I must say I wasn't aware of the subtle difference with tasks, but today I learned!

**Comment by fonsp** [comment]
It surprises me that internal Pluto code is showing up in the profile. We use `eval in boot.jl:360` to execute your code, but *your code* is doing the profile. So the profile is somehow going "up the call stack". Strange!

**Reply by gdalle** [reply]
I don't think it is so surprising. Since `@profile` works by periodically sampling backtraces during the call, there is no reason why it wouldn't go all the way up, right?

**Reply by fonsp** [reply]
Right, that explains it! Maybe this could be fixed in the profiling package: it can check the depth from which it is called, and discard everything below from the profile. This would get rid of the extra frames and fix your issue.

**Comment by karlwessel** [comment]
Did something change how cells are executed in Pluto? If I try to profile them I only get the `wait(); poptask()` part now. See https://github.com/pfitzseb/ProfileCanvas.jl/issues/37.
